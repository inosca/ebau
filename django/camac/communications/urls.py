from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"communications-topics", views.TopicView, "communications-topic")
r.register(r"communications-messages", views.MessageView, "communications-message")
r.register(
    r"communications-attachments", views.AttachmentView, "communications-attachment"
)

urlpatterns = r.urls
