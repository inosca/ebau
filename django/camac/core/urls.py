from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"publication-entries", views.PublicationEntryView, "publication")
r.register(r"authorities", views.AuthorityView, "authority")
r.register(r"workflow-entries", views.WorkflowEntryView, "workflow-entry")
r.register(r"resources", views.ResourceView, "resource")
r.register(r"instance-resources", views.InstanceResourceView, "instance-resource")
r.register(r"static-contents", views.StaticContentView, "static-content")
r.register(r"service-contents", views.ServiceContentViewSet, "service-content")

urlpatterns = r.urls
