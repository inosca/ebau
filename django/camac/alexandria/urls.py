from alexandria.core import views as alexandria_views
from django.urls import path
from rest_framework.routers import SimpleRouter

from camac.alexandria import views

r = SimpleRouter(trailing_slash=False)

# override the default viewset with our patched versions
# this needs to be updated when new urls are added in alexandria
r.register(r"categories", views.PatchedCategoryViewSet)
r.register(r"documents", views.PatchedDocumentViewSet)
r.register(r"files", views.PatchedFileViewSet)
r.register(r"tags", views.PatchedTagViewSet)
r.register(r"tagsynonymgroups", views.PatchedTagSynonymGroupViewSet)
r.register(r"marks", views.PatchedMarkViewSet)

urlpatterns = r.urls

urlpatterns.append(path("hook", alexandria_views.hook_view, name="hook"))
