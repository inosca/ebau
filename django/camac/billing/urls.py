from django.urls import re_path
from rest_framework.routers import SimpleRouter

from camac.billing import views
from camac.billing.export.views import BillingV2EntryExportView

urlpatterns = [
    re_path(
        r"billing-v2-entries/export$",
        BillingV2EntryExportView.as_view(),
        name="billing-export",
    ),
]

r = SimpleRouter(trailing_slash=False)

r.register(r"billing-v2-entries", views.BillingV2EntryViewset, "billing-v2-entry")

urlpatterns.extend(r.urls)
