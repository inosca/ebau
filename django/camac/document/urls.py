from django.urls import re_path
from rest_framework.routers import SimpleRouter

from . import views


class UnswaggeredAttachmentView(views.AttachmentView):
    swagger_schema = None


class UnswaggeredAttachmentSectionView(views.AttachmentSectionView):
    swagger_schema = None


r = SimpleRouter(trailing_slash=False)

r.register(r"attachments", views.AttachmentView)
r.register(r"attachment-sections", views.AttachmentSectionView)
r.register(r"templates", views.TemplateView)
r.register(r"attachment-download-history", views.AttachmentDownloadHistoryView)

urlpatterns = [
    re_path(
        r"^(?P<path>attachments/files/.+\..+)$",
        views.AttachmentDownloadView.as_view({"get": "retrieve"}),
        name="attachment-download",
    ),
    re_path(
        r"^attachments/files/",
        views.AttachmentDownloadView.as_view({"get": "list"}),
        name="multi-attachment-download",
    ),
]

urlpatterns.extend(r.urls)
