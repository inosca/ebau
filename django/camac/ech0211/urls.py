from django.urls import path, re_path
from django.views.generic import RedirectView

from camac.ech0211.views.kt_bern import (
    ApplicationsView as BEApplicationsView,
    ApplicationView as BEApplicationView,
    EventView as BEEventView,
    MessageView as BEMessageView,
    SendView as BESendView,
)
from camac.ech0211.views.kt_schwyz import (
    ApplicationsView as SZApplicationsView,
    ApplicationView as SZApplicationView,
)


class SZUrlsConf:
    urlpatterns = [
        re_path(
            r"application/(?P<instance_id>\d+)/?$",
            SZApplicationView.as_view({"get": "retrieve"}),
            name="application",
        ),
        re_path(
            r"applications",
            SZApplicationsView.as_view({"get": "list"}),
            name="applications",
        ),
    ]


class BEUrlsConf:
    redirects = {
        "instance/<int:instance_id>/": "/page/index/instance-resource-id/20074/instance-id/%(instance_id)i",
        "ebau-number/<int:instance_id>/": "/taskform/taskform/index/instance-resource-id/12000002/instance-id/%(instance_id)i",
        "claim/<int:instance_id>/": "/claim/claim/index/instance-resource-id/150000/instance-id/%(instance_id)i",
        "dossier-check/<int:instance_id>/": "/page/index/instance-resource-id/150009/instance-id/%(instance_id)i",
        "revision-history/<int:instance_id>/": "/revisionhistory/revisionhistory/index/instance-resource-id/150004/instance-id/%(instance_id)i",
    }

    urlpatterns = [
        re_path(
            r"application/(?P<instance_id>\d+)/?$",
            BEApplicationView.as_view({"get": "retrieve"}),
            name="application",
        ),
        re_path(
            r"applications",
            BEApplicationsView.as_view({"get": "list"}),
            name="applications",
        ),
        re_path(
            r"message/$", BEMessageView.as_view({"get": "retrieve"}), name="message"
        ),
        re_path(
            r"event/(?P<instance_id>(\d+))/(?P<event_type>(\w+))/?$",
            BEEventView.as_view({"post": "create"}),
            name="event",
        ),
        re_path(r"send/$", BESendView.as_view({"post": "create"}), name="send"),
    ] + [path(key, RedirectView.as_view(url=url)) for key, url in redirects.items()]
