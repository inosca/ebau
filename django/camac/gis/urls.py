from django.urls import re_path

from camac.gis import views

urlpatterns = [
    re_path(
        r"^data/(?:(?P<task_id>[\w-]+)/)?$",
        views.GISDataView.as_view(),
        name="gis-data",
    ),
    re_path(
        r"^apply$",
        views.GISApplyView.as_view(),
        name="gis-apply",
    ),
]
