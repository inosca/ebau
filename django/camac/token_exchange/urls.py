from django.urls import re_path

from camac.token_exchange.views import TokenExchangeView

urlpatterns = [
    re_path(
        r"^auth/token-exchange", TokenExchangeView.as_view(), name="token-exchange"
    ),
]
