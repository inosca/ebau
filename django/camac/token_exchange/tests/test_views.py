import json
import logging

import pytest
from django.urls import reverse
from pytest_lazy_fixtures import lf
from requests import HTTPError
from rest_framework import status


class FakeResponse:
    def __init__(self, data, *args, **kwargs):
        self.data = data

    def json(self):
        return self.data


@pytest.mark.freeze_time("2024-07-23 12:00:00")
@pytest.mark.parametrize(
    "token,keycloak_error,expected_status,error_message",
    [
        (None, False, status.HTTP_403_FORBIDDEN, None),
        (
            lf("invalid_jwt_token"),
            False,
            status.HTTP_403_FORBIDDEN,
            "InvalidSignature",
        ),
        (
            lf("jwt_token"),
            True,
            status.HTTP_403_FORBIDDEN,
            "some keycloak error",
        ),
        (lf("jwt_token"), False, status.HTTP_200_OK, None),
    ],
)
def test_token_exchange(
    db,
    caplog,
    clear_cache,
    error_message,
    expected_status,
    jwt_client,
    keycloak_error,
    mocker,
    snapshot,
    token,
):
    if keycloak_error:
        mocker.patch(
            "camac.token_exchange.keycloak.KeycloakClient.get_token",
            side_effect=HTTPError(
                response=FakeResponse({"errorMessage": "some keycloak error"})
            ),
        )
    else:
        mocker.patch("camac.token_exchange.keycloak.KeycloakClient.get_token")
        mocker.patch(
            "camac.token_exchange.keycloak.KeycloakClient.update_or_create_user"
        )
        mocker.patch(
            "camac.token_exchange.keycloak.KeycloakClient.token_exchange",
            return_value={"access_token": "my new access token"},
        )

    caplog.set_level(logging.ERROR)

    response = jwt_client.post(
        reverse("token-exchange"),
        data=json.dumps({"jwt-token": token}),
        content_type="application/json",
    )

    assert response.status_code == expected_status
    assert response.json() == snapshot

    if error_message:
        assert error_message in caplog.text


def test_token_exchange_token_reuse(db, clear_cache, jwt_client, jwt_token, mocker):
    mocker.patch("camac.token_exchange.keycloak.KeycloakClient.get_token")
    mocker.patch("camac.token_exchange.keycloak.KeycloakClient.update_or_create_user")
    mocker.patch(
        "camac.token_exchange.keycloak.KeycloakClient.token_exchange",
        return_value={"access_token": "my new access token"},
    )

    response1 = jwt_client.post(
        reverse("token-exchange"),
        data=json.dumps({"jwt-token": jwt_token}),
        content_type="application/json",
    )
    response2 = jwt_client.post(
        reverse("token-exchange"),
        data=json.dumps({"jwt-token": jwt_token}),
        content_type="application/json",
    )

    assert response1.status_code == status.HTTP_200_OK
    assert response2.status_code == status.HTTP_403_FORBIDDEN
    assert response2.json()["detail"] == "JWT token can't be used twice"
