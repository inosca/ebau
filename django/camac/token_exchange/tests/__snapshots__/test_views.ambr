# serializer version: 1
# name: test_token_exchange[None-False-403-None]
  dict({
    'detail': 'JWT token must be passed',
  })
# ---
# name: test_token_exchange[failed_decryption_jwt_token-False-403-InvalidSignature]
  dict({
    'detail': 'Invalid token',
  })
# ---
# name: test_token_exchange[invalid_exp_jwt_token-False-403-Expired at]
  dict({
    'detail': 'Invalid token',
  })
# ---
# name: test_token_exchange[invalid_iss_jwt_token-False-403-Invalid 'iss']
  dict({
    'detail': 'Invalid token',
  })
# ---
# name: test_token_exchange[invalid_signature_jwt_token-False-403-InvalidJWSSignature]
  dict({
    'detail': 'Invalid token',
  })
# ---
# name: test_token_exchange[jwt_token-False-200-None]
  dict({
    'access_token': 'my new access token',
  })
# ---
# name: test_token_exchange[jwt_token-True-403-some keycloak error]
  dict({
    'detail': 'Token could not be exchanged',
  })
# ---
